R/extract.R

```diff
@@ -45,7 +45,7 @@ extract.data.frame <- function(data, col, into, regex = "([[:alnum:]]+)",
 
   new_cols <- str_extract(value, into = into, regex = regex, convert = convert)
   out <- append_df(data, new_cols, var, remove = remove)
-  reconstruct_tibble(data, out, if (remove) var else chr())
+  df_cast(out, data, if (remove) var else chr())
```

Comment:
If you split `x` into `a`, `b`, `c` and `x` is grouped, wouldn't you want the result to be grouped by `a`, `b`, `c`?

Reply:
This makes sense but I'm not sure how we could express that generically.

Reply:
I was thinking that rather than using a vector of new variable names, you provide an integer vector that gives the source variable in the input for each variable in the output (using NA to indicate that it’s new)

Reply:
Right, but how could we express that through the vector interface of base or vctrs?

The only thing I can think of is to pass an ad hoc argument to `vec_cast()`. This may make sense, but since it's not generic and since that feature didn't previously exist, maybe we should wait for more use cases before introducing this approach.

Reply:
Yeah, we'd need to figure that out in the long term; but in tidyr it seems like we could just make it work through `[.grouped.df`, or some new generic specifically for this case? Maybe it would eventually become a standard argument for `vec_restore()` for container classes?

Reply:
I think I can make it work with only `[`, `vec_cbind()`, and `names<-`. But we need to add the restriction that splitted variables are a subtype of the original variable, which seems to make sense. This will also need fixes in other packages, including dplyr, because `names<-` is often buggy in instrumented data frame classes.

I will implement this new feature in a new PR.